File: packages/ui-flutter/lib/src/toaster.dart
import 'dart:async';

import 'package:flutter/material.dart';

import 'app_text.dart';
import 'theme.dart';

enum KinetixToastIntent { normal, success, error }

class KinetixToast {
  const KinetixToast(this.message, {this.intent = KinetixToastIntent.normal});

  final String message;
  final KinetixToastIntent intent;
}

/// Mirrors `packages/ui/src/components/sonner.tsx`. Place [KinetixToaster]
/// in a top-level `Stack` and drive it with a `KinetixToast?`; it shows a
/// bottom card and calls `onDismiss` after `duration`. The other ports
/// used the platform snackbar for the same role.
class KinetixToaster extends StatefulWidget {
  const KinetixToaster({
    super.key,
    required this.toast,
    required this.onDismiss,
    this.duration = const Duration(seconds: 3),
  });

  final KinetixToast? toast;
  final VoidCallback onDismiss;
  final Duration duration;

  @override
  State<KinetixToaster> createState() => _KinetixToasterState();
}

class _KinetixToasterState extends State<KinetixToaster> {
  Timer? _timer;

  @override
  void didUpdateWidget(covariant KinetixToaster oldWidget) {
    super.didUpdateWidget(oldWidget);
    final message = widget.toast?.message;
    if (message != null && message != oldWidget.toast?.message) {
      _timer?.cancel();
      _timer = Timer(widget.duration, widget.onDismiss);
    }
  }

  @override
  void dispose() {
    _timer?.cancel();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    final toast = widget.toast;
    if (toast == null) return const SizedBox.shrink();
    final c = KinetixTheme.of(context);
    final Color color = switch (toast.intent) {
      KinetixToastIntent.success => c.success,
      KinetixToastIntent.error => c.destructive,
      KinetixToastIntent.normal => c.popoverForeground,
    };

    return Align(
      alignment: Alignment.bottomCenter,
      child: Padding(
        padding: const EdgeInsets.all(16),
        child: Material(
          color: c.popover,
          borderRadius: BorderRadius.circular(8),
          elevation: 6,
          child: Container(
            constraints: const BoxConstraints(maxWidth: 400),
            padding: const EdgeInsets.symmetric(horizontal: 16, vertical: 12),
            decoration: BoxDecoration(
              borderRadius: BorderRadius.circular(8),
              border: Border.all(color: c.border, width: 1),
            ),
            child: Row(
              mainAxisSize: MainAxisSize.min,
              children: [
                if (toast.intent != KinetixToastIntent.normal) ...[
                  Icon(
                    toast.intent == KinetixToastIntent.success
                        ? Icons.check_circle_outline
                        : Icons.error_outline,
                    size: 16,
                    color: color,
                  ),
                  const SizedBox(width: 8),
                ],
                Flexible(child: Text(toast.message, style: AppText.bodyMd.copyWith(color: color))),
              ],
            ),
          ),
        ),
      ),
    );
  }
}

File: packages/ui-swiftui/Sources/KinetixUI/Toaster.swift
//
// Toaster.swift — KinetixToaster / KinetixToast.
//
// Mirrors packages/ui/src/components/sonner.tsx. Place KinetixToaster in
// an `.overlay { }` / top-level `ZStack` and drive it with a
// `KinetixToast?` binding; it shows a bottom card and clears itself after
// `duration` seconds. The Compose port used Material3's `SnackbarHost`
// for the same role.
//

import SwiftUI

public struct KinetixToast {
    public enum Intent {
        case normal, success, error
    }

    public let message: String
    public let intent: Intent

    public init(_ message: String, intent: Intent = .normal) {
        self.message = message
        self.intent = intent
    }
}

public struct KinetixToaster: View {
    @Environment(\.kinetixColors) private var colors

    @Binding private var toast: KinetixToast?
    private let duration: Double

    public init(toast: Binding<KinetixToast?>, duration: Double = 3) {
        self._toast = toast
        self.duration = duration
    }

    private func color(for intent: KinetixToast.Intent) -> Color {
        switch intent {
        case .normal:  return colors.popoverForeground
        case .success: return colors.success
        case .error:   return colors.destructive
        }
    }

    public var body: some View {
        VStack {
            Spacer()
            if let toast {
                HStack(spacing: 8) {
                    if toast.intent != .normal {
                        Image(systemName: toast.intent == .success ? "checkmark.circle" : "exclamationmark.circle")
                    }
                    Text(toast.message)
                        .font(.kinetixBody)
                    Spacer(minLength: 0)
                }
                .foregroundStyle(color(for: toast.intent))
                .padding(.horizontal, 16)
                .padding(.vertical, 12)
                .frame(maxWidth: 400)
                .background(colors.popover, in: RoundedRectangle(cornerRadius: 8, style: .continuous))
                .overlay {
                    RoundedRectangle(cornerRadius: 8, style: .continuous)
                        .strokeBorder(colors.border, lineWidth: 1)
                }
                .shadow(color: .black.opacity(0.15), radius: 12, y: 4)
                .padding(16)
                .transition(.move(edge: .bottom).combined(with: .opacity))
                .task(id: toast.message) {
                    try? await Task.sleep(nanoseconds: UInt64(duration * 1_000_000_000))
                    withAnimation { self.toast = nil }
                }
            }
        }
    }
}
